R/formula.R

```diff
@@ -0,0 +1,61 @@
+#' Build formula for `mgcv::gam()`
+#'
+#' Build up the formula as a string and and return a formula object meant for
+#' use by [`mgcv::gam()`]. The formula components are built up based on the
+#' parameters passed to [`RtGam`]. The model makes the assumptions that the
+#' smoothness of epidemic dynamics does not change dramatically over a three
+#' week period. For periods of three weeks or shorter, the model uses a simple
+#' thin plate spline for the global epidemic trend. For periods of longer than
+#' three days, the model uses an adaptive smoother with an additional penalty
+#' basis for each additional three week period. The model smoothly interpolates
+#' between the penalty bases and uses p-splines for the smoothing basis.
+#'
+#' Currently support for groups via hierarchical modeling  is not supported,
+#' but when implemented it will use Model GS from Pederson et al., 2019.
+#'
+#' @param n_timesteps Number of distinct timesteps in the dataframe
+#'  returned from [`prepare_inputs()`]
+#' @param k Global basis dimension to be partitioned between the model smooths
+#' @param m Penalty basis dimension on the global smooth
+#' @param is_grouped Whether to use a hierarchical model. Not yet supported.
+#' @return A formula to be used by [`mgcv::gam()`]
+#' @noRd
+formula_creator <- function(k, m, is_grouped) {
+  outcome <- "cases"
+  intercept <- "1"
+
+  smooth_basis_dim <- smooth_basis_creator(k)
+
+  # Apply adaptive spline if 3 weeks or more of data are available
+  # nolint start
+  if (m > 1) {
+    # With adaptive basis, m refers to the order of the penalty matrix not the
+    # order of the smoothing penalty as it does in the other smoothing bases.
+    plus_global_trend <- glue::glue("+ s(timesteps,
+                                         k = {smooth_basis_dim[['global_trend']]},
+                                         m = {m},
+                                         bs = 'ad')") # nolint
+  } else {
+    # Adaptive penalty with `m = 1` is equivalent to a non-adaptive smooth but
+    # thin-plate performance is a bit better than p-spline, so preference to
+    # fall back to thin-plate.
+    plus_global_trend <- glue::glue("+ s(timesteps,
```

Comment:
This is a very boring comment. 

It might read more naturally if the plus here was in line 48. This would also slightly reduce copy and paste.

Reply:
Ah yes, I thought about this when I was designing this. The idea is that we want to implement optional local trends and day of week effects. So the formula will be something like

`{outcome} ~ {intercept} {plus_global_trend} {plus_local_trend} {plus_day_of_week}`

If instead it were

`{outcome} ~ {intercept} + {global_trend} + {local_trend} + {day_of_week}` 

and we take out the local trend or day of week then the dangling `+` would cause the formula parsing to error. By assigning the `+` in the variable to be evaluated we can mix and match as needed.

Reply:
I see. that is quite a compelling arg. My counter arg is can you just on the fly construct a string with + between the parts? that seems like the better option.

Reply:
Above would reduce the + spam quite a bit once there are multiple formula components in play

Reply:
You mean like this?

```r
global_trend <- "s(x)"
local_trend <- "s(y)"
day_of_week <- NULL

formula <- paste("outcome ~ 1", global_trend, local_trend, day_of_week, sep = "+")
```

That still doesn't work because the last NULL is still evaluated so it returns

>  "outcome ~ 1+s(x)+s(y)+"

Reply:
I did mean like this but I hadn't thought about the null handling. You could sidestep it by first composing what will be combined (and drop NULLS) but it starts to get less clear. I think leave as is.